lib/charms/haproxy/v0/haproxy_route.py

```diff
@@ -0,0 +1,1191 @@
+            rate_limit_connections_per_minute: Maximum connections allowed per minute.
+            rate_limit_policy: Policy to apply when rate limit is reached.
+            upload_limit: Maximum upload bandwidth in bytes per second.
+            download_limit: Maximum download bandwidth in bytes per second.
+            retry_count: Number of times to retry failed requests.
+            retry_interval: Interval between retries in seconds.
+            retry_redispatch: Whether to redispatch failed requests to another server.
+            deny_paths: List of paths that should not be routed to the backend.
+            server_timeout: Timeout for requests from haproxy to backend servers in seconds.
+            client_timeout: Timeout for client requests to haproxy in seconds.
+            queue_timeout: Timeout for requests waiting in queue in seconds.
+            server_maxconn: Maximum connections per server.
+            host: Hostname or IP address of the unit (if not provided, will use binding address).
+        """
+        super().__init__(charm, relation_name)
+
+        self._relation_name = relation_name
+        self.relation = self.model.get_relation(self._relation_name)
+        self.charm = charm
+        self.app = self.charm.app
+
+        # build the full application data
+        self._application_data = self._generate_application_data(
+            service,
+            ports,
+            paths,
+            subdomains,
+            check_interval,
+            check_rise,
+            check_fall,
+            check_paths,
+            path_rewrite_expressions,
+            query_rewrite_expressions,
+            header_rewrite_expressions,
+            load_balancing_algorithm,
+            load_balancing_cookie,
+            rate_limit_connections_per_minute,
+            rate_limit_policy,
+            upload_limit,
+            download_limit,
+            retry_count,
+            retry_interval,
+            retry_redispatch,
+            deny_paths,
+            server_timeout,
+            client_timeout,
+            queue_timeout,
+            server_maxconn,
+        )
+        self._host = host
+
+        on = self.charm.on
+        self.framework.observe(on[self._relation_name].relation_created, self._configure)
+        self.framework.observe(on[self._relation_name].relation_changed, self._configure)
+        self.framework.observe(on[self._relation_name].relation_broken, self._on_relation_broken)
+
+    def _configure(self, _: EventBase) -> None:
+        """Handle relation events."""
+        self.update_relation_data()
+        if self.relation and self.get_proxied_endpoints():
+            self.on.ready.emit()
+
+    def _on_relation_broken(self, _: RelationBrokenEvent) -> None:
+        """Handle relation broken event."""
+        self.on.removed.emit()
+
+    # pylint: disable=too-many-arguments,too-many-positional-arguments,too-many-locals
+    def provide_haproxy_route_requirements(
```

Comment:
We should set service and ports as mandatory here, as we want to inform the dev using the library that something is missing (so they can act or set the appropriate status)